platform/inc/can.hpp

```diff
@@ -0,0 +1,230 @@
+    uint8_t message_marker;
+    uint8_t* data;
+    uint32_t data_length;
+};
+
+struct RxCanMessage : public CanMessage {
+    RxCanMessage(uint8_t *data, uint32_t data_length=MAX_DATA_LENGTH);
+    uint32_t filter_index=0;
+
+};
+
+struct CanDriver {
+
+    enum class OperatingMode : uint32_t {
+        Normal = FDCAN_MODE_NORMAL,
+        RestrictedOperation,
+        BusMonitoring,
+        InternalLoopback,
+        ExternalLoopback
+    };
+
+
+    /**
+     * @brief Initalized the Can Driver
+     * This function should only be called once in the setup of the
+     * application.
+     *
+     */
+    void initialize(OperatingMode initial_operating_mode=OperatingMode::InternalLoopback);
+
+    /**
+     * @brief Test the CAN Driver
+     * A diagnostic tool for testing the can driver.
+     * Disconnects the Device from the CAN bus, runs some tests,
+     * and then places the device back on the CAN bus.
+     */
+    void test_driver();
+
+    /**
+     * @brief Get the Single Instance of the Can Driver
+     * The CAN Driver is designed with a singleton pattern due
+     * to it's association with physical hardware.
+     *
+     * @return CanDriver*
+     */
+    static CanDriver &get_driver() {
+        static CanDriver can_driver;
+        return can_driver;
+    }
+
+    /**
+     * @brief Get the operating mode
+     * Determine the current operating mode of the CAN Driver
+     *
+     * @return OperatingMode
+     */
+    OperatingMode get_operating_mode() const;
+
+    /**
+     * @brief Set the Operating Mode of the can bus
+     * takes the device off of the can bus, switches operating modes to the mode
+     * specified, then places the device back on the can bus.
+     *
+     * @param new_operating_mode
+     * @return true
+     * @return false
+     */
+    [[nodiscard]] bool set_operating_mode(OperatingMode new_operating_mode);
+
+    /**
+     * @brief A non blocking write to the CAN bus
+     *
+     * @return uint32_t
+     * The returned txId can be used with await_write to
+     * block until the message is sent on the bus.
+     */
+    uint32_t write(CanMessage &msg);
+
+    /**
+     * @brief Block until the message specified by txId
+     * is sent.
+     *
+     * @param txId
+     */
+    void await_write(uint32_t &txId);
+
+    [[nodiscard]] bool rx_ready(CanRxFifo rxFifo = DEFAULT_RX_FIFO);
+
+    [[nodiscard]] bool read(RxCanMessage &msg, CanRxFifo rxFifo = DEFAULT_RX_FIFO);
+
+
+    template<typename... Filter>
+    [[nodiscard]] bool push_filters(Filter... filter) {
+        static_assert(is_type<CanMessageFilter, Filter...>(),
+        "Filter should be of type CanMessageFilter");
+        auto result = false;
+        auto status = HAL_FDCAN_Stop(&can_handle);
+        if (status != HAL_OK) { Error_Handler(); }
+        result = (push_filter(filter) && ...);
+        status = HAL_FDCAN_Start(&can_handle);
+        if (status != HAL_OK) { Error_Handler(); }
+
+        return result;
+    }
```

Comment:
Just a Note, this function is the reason that I had to bump to c++17. I believe we wanted to stay away from c++20 which is the reason that static_assert is there (there is a cleaner wait to do this in cpp20, but this isn't too bad either).